clayui.com/src/components/IconSearch/index.js

```diff
@@ -0,0 +1,84 @@
+/**
+ * © 2020 Liferay, Inc. <https://liferay.com>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import ClayForm, {ClayInput} from '@clayui/form';
+import ClayIcon from '@clayui/icon';
+import React, {useMemo, useState} from 'react';
+
+const spritemap = '/images/icons/icons.svg';
+
+/**
+ * Component that allows searching through icons
+ * @param {string} label - Label of the search input field.
+ * @param {string} placeholder - Placeholder of the search input field.
+ * @param {Array[Object]} source - Path to the source JSON. Needs to follow pattern of [{name: String, aliases: Array<String>}]
+ * @param {string} type - Type of icons being provided, supports "icons" and "flags"
+ */
+const IconSearch = ({
+	label = 'Search Icons',
+	placeholder = 'Search Icons...',
+	source,
+	type = 'icons',
+}) => {
+	const [searchQuery, setSearchQuery] = useState('');
+	let list = [];
+
+	const filteredIcons = useMemo(() => {
+		const query = searchQuery.toLowerCase();
+
+		return source.filter(
+			({aliases, name}) =>
+				name.toLowerCase().includes(query) ||
+				aliases.some(alias => alias.toLowerCase().includes(query))
+		);
+	}, [searchQuery]);
+
+	if (filteredIcons.length) {
+		list = filteredIcons;
+	} else {
+		list = searchQuery ? [] : source;
+	}
+
+	return (
+		<>
+			<ClayForm.Group>
+				<label style={{width: '100%'}}>
+					{label}
+
+					<ClayInput
+						onChange={event => setSearchQuery(event.target.value)}
+						placeholder={placeholder}
+						type="text"
+						value={searchQuery}
+					/>
+				</label>
+
+				<ul className="lexicon-icon-list list-unstyled">
+					{list.map(icon => (
+						<li key={icon.name}>
+							<ClayIcon
+								spritemap={spritemap}
+								symbol={icon.name}
+							/>
+
+							<span>
+								{type === 'flags'
```

Comment:
Rather than creating a one off API for a specific case like flags, let's try to make the API as generic and configurable as possible.

For example, instead of a `type` prop, we could do something like `iconLabelFormatter` which is a function that renders the label for the icon. If we did this, we can pass it in as a prop for flags and then internally have a default function for it.

```js
const IconSearch = ({
	label = 'Search Icons',
	iconLabelFormatter = (item) => item.name,
	placeholder = 'Search Icons...',
	source,
}) => {
//...
<li key={icon.name}>
	<ClayIcon
		spritemap={spritemap}
		symbol={icon.name}
	/>

	<span>
		{iconLabelFormatter(icon)}
	</span>
</li>
//...
}
```



_Side note / soap box_
**While this may not actually be a huge issue for us on the clayui.com codebase or impact us greatly, the goal would be to get in the habit of designing more generic APIs.** Otherwise, we may start designing APIs in our components that require us to add additional "cases" and logic for different circumstances. For example, suppose we have the following data and an additional case where we need to render the Spanish word for each icon.

```jsx
// aliases: [english, spanish, french]
[
	{
		name: 'icon-apple',
		aliases: ['apple', 'manzana', 'pomme'],
	},
	{
		name: 'icon-banana',
		aliases: ['banana', 'plátano', 'banane'],
	},
];

// With the existing API, we would have to add a "spanish" type and add the logic to
// handle that case.
<IconSearch type="spanish" source={icons} />

// If we go with a more generic API, we won't have to add any internal logic or types, and
// it can be used as is.
<IconSearch iconLabelFormatter={(item) => item.aliased[1]} source={icons} />
```

By adopting a more generic API, we are making it easier for ourselves to maintain over time and avoiding having to create a new API for every case.

This whole idea sort of relates to [big-o notation](https://rob-bell.net/2009/06/a-beginners-guide-to-big-o-notation/) and complexity. That link may help give more of a theoretical idea of complexity issues, but if it creates confusion, feel free to ignore it. 
/end

I hope that example is helpful for clarity on API design. And as always, if you have any questions just let me know or I would be happy to get on a call and talk through it more.

Reply:
You're right. I think that if we spent more time defining the requirements of this component before going into writing the implementation we would have avoided this problem.

This did cross my mind, and it did feel bad to hardcode the type, but I didn't go far in my thinking to get the idea of giving the user of the component the option to pass a formatting function. I'll have to think about that stuff in the future!